### PR TITLE
planner/core: fix index out of bound bug when empty dual table is remove for mpp query

### LINKED ISSUE #28250: query failed with `index out of range [-1]`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> drop table if exists t1;
Query OK, 0 rows affected (0.51 sec)

mysql> drop table if exists t;
Query OK, 0 rows affected (0.52 sec)

mysql> create table t (a int not null, b int, c varchar(20));
Query OK, 0 rows affected (0.51 sec)

mysql> create table t1 (a int, b int not null, c double);
Query OK, 0 rows affected (0.51 sec)

mysql> alter table t set tiflash replica 1;
Query OK, 0 rows affected (0.51 sec)

mysql> alter table t1 set tiflash replica 1;
Query OK, 0 rows affected (0.51 sec)

mysql> explain select count(*) from (select a , b from t union all select a , c from t1 where false) tt;
+------------------------------------+----------+-------------------+---------------+-----------------------------------------------------+
| id                                 | estRows  | task              | access object | operator info                                       |
+------------------------------------+----------+-------------------+---------------+-----------------------------------------------------+
| HashAgg_33                         | 1.00     | root              |               | funcs:count(Column#14)->Column#11                   |
| └─TableReader_35                   | 1.00     | root              |               | data:ExchangeSender_34                              |
|   └─ExchangeSender_34              | 1.00     | batchCop[tiflash] |               | ExchangeType: PassThrough                           |
|     └─HashAgg_15                   | 1.00     | batchCop[tiflash] |               | funcs:count(1)->Column#14                           |
|       └─Union_32                   | 10000.00 | batchCop[tiflash] |               |                                                     |
|         └─Projection_51            | 10000.00 | batchCop[tiflash] |               | cast(Column#9, int(11) BINARY)->Column#9, Column#10 |
|           └─Projection_26          | 10000.00 | batchCop[tiflash] |               | test.t.a, cast(test.t.b, double BINARY)->Column#10  |
|             └─TableFullScan_27     | 10000.00 | batchCop[tiflash] | table:t       | keep order:false, stats:pseudo                      |
+------------------------------------+----------+-------------------+---------------+-----------------------------------------------------+
8 rows in set (0.00 sec)

mysql>  select count(*) from (select a , b from t union all select a , c from t1 where false) tt;
ERROR 1105 (HY000): runtime error: index out of range [-1]
```
### 2. What did you expect to see? (Required)
query executed without error
### 3. What did you see instead (Required)
ERROR 1105 (HY000): runtime error: index out of range [-1]
### 4. What is your TiDB version? (Required)
master@38e90ad13657ebe2cd2e2a2832aba26130b75e64

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #28250 

Problem Summary:
As described in the issue.
### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:
The root cause is after #25218, union may contain less than 2 children, but in `stringer.ToString`, it still assume that `union` always contains at least 2 children. This pr refine the related code to consider the case that `union` contains less than 2 children.

### Check List 

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Fix index out of bound bug when empty dual table is remove for mpp query
```

### CODE REVIEW

**Comment by windtalker** [comment]
/needs-cherry-pick-5.0

**Review by wjhuang2016 — APPROVED** [review]

File: planner/core/stringer.go [hunk 1]
```diff
@@ -28,10 +28,25 @@ func ToString(p Plan) string {
 	return strings.Join(strs, "->")
 }
 
+func needIncludeChildrenString(plan Plan) bool {
+	switch x := plan.(type) {
+	case *LogicalUnionAll, *PhysicalUnionAll, *LogicalPartitionUnionAll:
+		// after https://github.com/pingcap/tidb/pull/25218, the union may contain less than 2 children,
+		// but we still wants to include its child plan's information when calling `toString` on union.
+		return true
```
Comment (fzhedu) [inline planner/core/stringer.go:36]: would this case happen in other cases？

Comment (windtalker) [reply planner/core/stringer.go:36]: You mean will `the union may contain less than 2 children` cause other bugs? I'm not sure about this, and if it does happen, we will fix it in other prs.

Comment (tiancaiamao) [reply planner/core/stringer.go:36]: oncall 3751 can't reproduce, but maybe it's caused by the same bug? @windtalker

**Review by fzhedu — APPROVED** [review]
/LGTM

**Comment by fzhedu** [comment]
Reducing the dual table child of union may introduce other unknown issues, we should remove union if its children are less than 2.

**Pull request merged**